File: dbflow/src/main/java/com/raizlabs/android/dbflow/sql/saveable/ListModelSaver.java
package com.raizlabs.android.dbflow.sql.saveable;

import android.support.annotation.NonNull;

import com.raizlabs.android.dbflow.structure.database.DatabaseStatement;
import com.raizlabs.android.dbflow.structure.database.DatabaseWrapper;

import java.util.Collection;

public class ListModelSaver<TModel> {

    private final ModelSaver<TModel> modelSaver;

    public ListModelSaver(ModelSaver<TModel> modelSaver) {
        this.modelSaver = modelSaver;
    }

    public synchronized void saveAll(@NonNull Collection<TModel> tableCollection) {
        saveAll(tableCollection, modelSaver.getWritableDatabase());
    }

    public synchronized void saveAll(@NonNull Collection<TModel> tableCollection,
                                     @NonNull DatabaseWrapper wrapper) {
        // skip if empty.
        if (tableCollection.isEmpty()) {
            return;
        }

        DatabaseStatement statement = modelSaver.getModelAdapter().getInsertStatement(wrapper);
        DatabaseStatement updateStatement = modelSaver.getModelAdapter().getUpdateStatement(wrapper);
        try {
            for (TModel model : tableCollection) {
                modelSaver.save(model, wrapper, statement, updateStatement);
            }
        } finally {
            statement.close();
        }
    }

    public synchronized void insertAll(@NonNull Collection<TModel> tableCollection) {
        insertAll(tableCollection, modelSaver.getWritableDatabase());
    }

    public synchronized void insertAll(@NonNull Collection<TModel> tableCollection,
                                       @NonNull DatabaseWrapper wrapper) {
        // skip if empty.
        if (tableCollection.isEmpty()) {
            return;
        }

        DatabaseStatement statement = modelSaver.getModelAdapter().getInsertStatement(wrapper);
        try {
            for (TModel model : tableCollection) {
                modelSaver.insert(model, statement, wrapper);
            }
        } finally {
            statement.close();
        }
    }

    public synchronized void updateAll(@NonNull Collection<TModel> tableCollection) {
        updateAll(tableCollection, modelSaver.getWritableDatabase());
    }

    public synchronized void updateAll(@NonNull Collection<TModel> tableCollection,
                                       @NonNull DatabaseWrapper wrapper) {
        // skip if empty.
        if (tableCollection.isEmpty()) {
            return;
        }

        DatabaseStatement updateStatement = modelSaver.getModelAdapter().getUpdateStatement(wrapper);
        try {
            for (TModel model : tableCollection) {
                modelSaver.update(model, wrapper, updateStatement);
            }
        } finally {
            updateStatement.close();
        }
    }

    public synchronized void deleteAll(@NonNull Collection<TModel> tableCollection) {
        deleteAll(tableCollection, modelSaver.getWritableDatabase());
    }

    public synchronized void deleteAll(@NonNull Collection<TModel> tableCollection,
                                       @NonNull DatabaseWrapper wrapper) {
        // skip if empty.
        if (tableCollection.isEmpty()) {
            return;
        }

        DatabaseStatement deleteStatement = modelSaver.getModelAdapter().getDeleteStatement(wrapper);
        try {
            for (TModel model : tableCollection) {
                modelSaver.delete(model, deleteStatement, wrapper);
            }
        } finally {
            deleteStatement.close();
        }
    }

    @NonNull
    public ModelSaver<TModel> getModelSaver() {
        return modelSaver;
    }
}

File: dbflow/src/main/java/com/raizlabs/android/dbflow/sql/saveable/ListModelSaver.kt
package com.raizlabs.android.dbflow.sql.saveable

import com.raizlabs.android.dbflow.structure.database.DatabaseStatement
import com.raizlabs.android.dbflow.structure.database.DatabaseWrapper

open class ListModelSaver<T : Any>(val modelSaver: ModelSaver<T>) {

    val modelAdapter = modelSaver.modelAdapter

    @Synchronized
    open fun saveAll(tableCollection: Collection<T>,
                     wrapper: DatabaseWrapper): Long {
        val statement = modelAdapter.getInsertStatement(wrapper)
        val updateStatement = modelAdapter.getUpdateStatement(wrapper)
        return applyAndCount(tableCollection, statement, updateStatement) {
            modelSaver.save(it, statement, updateStatement, wrapper)
        }
    }

    @Synchronized
    open fun insertAll(tableCollection: Collection<T>,
                       wrapper: DatabaseWrapper): Long {
        val statement = modelAdapter.getInsertStatement(wrapper)
        return applyAndCount(tableCollection, statement) { modelSaver.insert(it, statement, wrapper) > 0 }
    }

    @Synchronized
    open fun updateAll(tableCollection: Collection<T>,
                       wrapper: DatabaseWrapper): Long {
        val statement = modelAdapter.getUpdateStatement(wrapper)
        return applyAndCount(tableCollection, statement) { modelSaver.update(it, statement, wrapper) }
    }

    @Synchronized
    open fun deleteAll(tableCollection: Collection<T>,
                       wrapper: DatabaseWrapper): Long {
        val statement = modelAdapter.getDeleteStatement(wrapper)
        return applyAndCount(tableCollection, statement) { modelSaver.delete(it, statement, wrapper) }
    }

    private inline fun applyAndCount(tableCollection: Collection<T>,
                                     databaseStatement: DatabaseStatement,
                                     otherStatement: DatabaseStatement? = null,
                                     crossinline fn: (T) -> Boolean): Long {
        // skip if empty.
        if (tableCollection.isEmpty()) {
            return 0L
        }

        var count = 0L
        try {
            tableCollection.forEach {
                if (fn(it)) {
                    count++
                }
            }
        } finally {
            databaseStatement.close()
            otherStatement?.close()
        }
        return count
    }
}
